### PR TITLE
feat(mcp): add MCP-009 to MCP-012 server config validation rules

### LINKED ISSUE #298: New MCP rules: MCP-009 to MCP-012 (.mcp.json server type validation)
## Summary

Add 4 new validation rules for `.mcp.json` server configuration entries.

### MCP-009: Missing command for stdio server (HIGH)
When a server entry has `type: "stdio"` (or no type, since stdio is the default), the `command` field is required. Without it the server cannot start.

### MCP-010: Missing url for http/sse server (HIGH)
When a server entry has `type: "http"` or `type: "sse"`, the `url` field is required.

### MCP-011: Invalid MCP server type (HIGH)
The `type` field must be `stdio`, `http`, or `sse`. Any other value is invalid.

### MCP-012: Deprecated SSE transport (MEDIUM)
SSE transport is deprecated in favor of Streamable HTTP. Suggest changing `type: "sse"` to `type: "http"` when the server supports it.

## Context

Current MCP rules (MCP-001 to MCP-008) validate JSON-RPC version, tool fields, inputSchema, description length, consent mechanisms, annotations, parse errors, and protocol version. But the `.mcp.json` server entry schema itself (type, command, url, args, env) has no validation.

## Implementation

- Extend `McpValidator` in `crates/agnix-core/src/rules/mcp.rs`
- Add rules to `knowledge-base/rules.json` and `VALIDATION-RULES.md`
- Add test fixtures in `tests/fixtures/mcp/`

## References

- [MCP Specification](https://modelcontextprotocol.io/specification)
- `agent-knowledge/agent-config-optional-fields.md`

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Adds 4 new MCP validation rules for `.mcp.json` server configuration entries, extending the existing MCP-001 to MCP-008 suite.

Closes #298

## New Rules

- **MCP-009**: Missing `command` for stdio server -- when `type: "stdio"` (or no type, since stdio is the default), the `command` field is required
- **MCP-010**: Missing `url` for http/sse server -- when `type: "http"` or `type: "sse"`, the `url` field is required
- **MCP-011**: Invalid MCP server type -- `type` must be `stdio`, `http`, or `sse`
- **MCP-012**: Deprecated SSE transport -- warns that SSE is deprecated in favor of Streamable HTTP, with an unsafe auto-fix to change `"sse"` to `"http"`

## Changes

- `crates/agnix-core/src/rules/mcp.rs` -- new `validate_server()` function with MCP-009 to MCP-012 logic
- `crates/agnix-core/src/schemas/mcp.rs` -- `McpServerConfig` struct, `McpConfigSchema` with `mcpServers` field, `VALID_MCP_SERVER_TYPES` constant
- `knowledge-base/rules.json` -- 4 new rule entries with full evidence metadata
- `knowledge-base/VALIDATION-RULES.md` -- human-readable docs for the 4 rules
- `locales/en.yml` -- i18n message strings for MCP-009 to MCP-012
- `tests/fixtures/mcp/` -- 7 new fixture files for server config validation
- `website/docs/rules/generated/` -- generated docs for MCP-009 to MCP-012
- `crates/agnix-cli/src/sarif.rs` -- updated rule count
- `crates/agnix-cli/tests/rule_parity.rs` -- updated parity tests
- `crates/agnix-mcp/tests/mcp_tests.rs` -- updated rule count assertion (102 -> 106)

## Test plan

- [x] All 1835+ tests pass (including 30+ new MCP server config tests)
- [x] Release build succeeds
- [x] CLAUDE.md and AGENTS.md are byte-for-byte identical
- [x] Rule parity tests pass (rules.json <-> VALIDATION-RULES.md)
- [x] Each new rule can be individually disabled via config
- [x] Invalid type (MCP-011) short-circuits to avoid false MCP-009/MCP-010 errors
- [x] MCP-012 auto-fix correctly targets `"sse"` -> `"http"` replacement

### CODE REVIEW

File: crates/agnix-core/src/rules/mcp.rs [hunk 1]
```diff
@@ -578,6 +585,109 @@ fn validate_tool(
...
+                    "MCP-009",
+                    t!("rules.mcp_009.message", server = name),
+                )
+                .with_suggestion(t!("rules.mcp_009.suggestion")),
+            );
+        }
+    }
+
+    // MCP-010: Missing url for http/sse server
+    if config.is_rule_enabled("MCP-010") && (effective_type == "http" || effective_type == "sse") {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs]: Applied - changed to `["http", "sse"].contains(&effective_type)` in 7e19eed.

File: website/docs/rules/generated/mcp-009.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  "jsonrpc": "2.0",
+  "tools": [
+    {
+      "name": "validate_file",
+      "description": "Validate one configuration file",
+      "inputSchema": {"type": "object"}
+    }
+  ]
+}
+```
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply website/docs/rules/generated/mcp-009.md:60]: The generated docs come from a script that uses generic examples per rule category. Updating the generation logic is out of scope for this PR but would be a good follow-up.

File: crates/agnix-core/src/schemas/mcp.rs [hunk 3]
```diff
@@ -50,10 +50,17 @@ pub struct McpJsonRpcMessage {
...
+/// Valid MCP server transport types
+pub const VALID_MCP_SERVER_TYPES: &[&str] = &["stdio", "http", "sse"];
+
 /// MCP server configuration (as used in .mcp.json or settings.json)
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct McpServerConfig {
-    /// Command to run the server
+    /// Transport type: stdio (default), http, or sse
+    #[serde(rename = "type")]
+    pub server_type: Option<String>,
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/mcp.rs:61]: Valid point. When type is a non-string value like \, serde's \ deserialization returns \, causing the whole \ fallback. This is a pre-existing limitation of the deserialization approach. Filing as a follow-up since fixing it requires a custom deserializer or \ fallback for the type field.

File: crates/agnix-core/src/rules/mcp.rs [hunk 4]
```diff
@@ -578,6 +585,109 @@ fn validate_tool(
...
+    config: &LintConfig,
+    diagnostics: &mut Vec<Diagnostic>,
+) {
+    let (line, col) = find_json_field_location(content, name);
+
+    // Determine effective type (default is "stdio" when type is absent)
+    let effective_type = server
+        .server_type
+        .as_deref()
+        .unwrap_or("stdio");
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs]: Acknowledged. The field locator is a known limitation across the MCP validator (used for all MCP-001 through MCP-012). Scoping the search to within the mcpServers object would be better but is a cross-cutting improvement, not specific to this PR.

File: crates/agnix-core/src/schemas/mcp.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/mcp.rs:61]: Valid point. When type is a non-string value like 123, serde deserialization of Option<String> returns Err, causing the whole McpConfigSchema fallback. This is a pre-existing limitation of the deserialization approach. Filing as a follow-up since fixing it requires a custom deserializer or Value fallback for the type field.

File: crates/agnix-core/src/rules/mcp.rs [hunk 5]
```diff
@@ -578,6 +585,109 @@ fn validate_tool(
...
+                Diagnostic::error(
+                    path.to_path_buf(),
+                    line,
+                    col,
+                    "MCP-009",
+                    t!("rules.mcp_009.message", server = name),
+                )
+                .with_suggestion(t!("rules.mcp_009.suggestion")),
+            );
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs:646]: Fixed in 7e19eed. MCP-009 now treats null, empty string, and empty array commands as missing. Added 3 new tests for these edge cases.

File: crates/agnix-core/src/rules/mcp.rs [hunk 6]
```diff
@@ -578,6 +585,109 @@ fn validate_tool(
...
+                    t!(
+                        "rules.mcp_010.message",
+                        server = name,
+                        server_type = effective_type
+                    ),
+                )
+                .with_suggestion(t!("rules.mcp_010.suggestion")),
+            );
+        }
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs:669]: Fixed in 7e19eed. MCP-010 now treats empty and whitespace-only URLs as missing. Added 2 new tests for these edge cases.

File: crates/agnix-core/src/rules/mcp.rs [hunk 7]
```diff
@@ -578,6 +585,109 @@ fn validate_tool(
...
+
+        // Unsafe auto-fix: change "sse" to "http"
+        if let Some((start, end)) = find_unique_json_string_value_span(content, "type", "sse") {
+            diag = diag.with_fix(Fix::replace(
+                start,
+                end,
+                "http",
+                t!("rules.mcp_012.fix"),
+                false,
+            ));
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs:690]: Known limitation. The uniqueness constraint means only single-SSE-server files get the auto-fix. Multi-server files still get the warning, just without the fix.

File: knowledge-base/VALIDATION-RULES.md [hunk 8]
```diff
@@ -1051,4 +1081,4 @@ pub fn validate_skill(path: &Path, content: &str) -> Vec<Diagnostic> {
 **Total Coverage**: 102 validation rules across 15 categories
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/VALIDATION-RULES.md]: Fixed in 01b3717. All doc references now show 106.

File: crates/agnix-core/src/rules/mcp.rs [hunk 9]
```diff
@@ -578,6 +585,106 @@ fn validate_tool(
...
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs:691]: Duplicate of the earlier auto-fix uniqueness comment. See reply above.

File: crates/agnix-core/src/rules/mcp.rs [hunk 10]
```diff
@@ -170,6 +170,13 @@ impl Validator for McpValidator {
             validate_tool(tool, path, content, config, &mut diagnostics, idx);
         }
 
+        // Validate MCP server configurations (MCP-009 to MCP-012)
+        if let Some(servers) = &mcp_config.mcp_servers {
+            for (name, server) in servers {
+                validate_server(name, server, path, content, config, &mut diagnostics);
+            }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/mcp.rs:177]: This is the same as the Codex concern about non-string type values causing deserialization failure. Acknowledged as a pre-existing limitation. See reply to the Codex comment.

File: knowledge-base/INDEX.md [hunk 11]
```diff
@@ -92,14 +92,14 @@ knowledge-base/
...
-| MCP | 8 | 7 | 1 | 0 | 1 |
+| MCP | 12 | 10 | 2 | 0 | 2 |
 | XML | 3 | 3 | 0 | 0 | 1 |
 | References | 2 | 2 | 0 | 0 | 0 |
 | Prompt Eng | 4 | 0 | 4 | 0 | 1 |
 | Cross-Platform | 6 | 4 | 2 | 0 | 0 |
 | Cursor | 6 | 3 | 3 | 0 | 1 |
 | Version Awareness | 1 | 0 | 0 | 1 | 0 |
-| **TOTAL** | **108** | **81** | **26** | **1** | **22** |
+| **TOTAL** | **112** | **84** | **27** | **1** | **22** |
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/INDEX.md]: Fixed. Updated INDEX.md MCP row Auto-Fix from 2 to 3 and TOTAL from 22 to 23.
